src/desktop/input_capture.rs

```diff
@@ -1,3 +1,269 @@
+//!                 Position::Top => (x, y, x + width - 1, y),
+//!                 Position::Bottom => (x, y + height, x + width - 1, y + height),
+//!             };
+//!             Barrier::new(id, barrier_pos)
+//!         })
+//!         .collect::<Vec<_>>();
+//!
+//!     eprintln!("requested barriers: {barriers:?}");
+//!
+//!     let request = input_capture
+//!         .set_pointer_barriers(&session, &barriers, zones.zone_set())
+//!         .await?;
+//!     let response = request.response()?;
+//!     let failed_barrier_ids = response.failed_barriers();
+//!
+//!     eprintln!("failed barrier ids: {:?}", failed_barrier_ids);
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! ## Enabling Input Capture and Retrieving Captured Input Events.
+//!
+//! The following full example uses the [reis crate](https://docs.rs/reis/0.2.0/reis/)
+//! for libei communication.
+//!
+//! Input Capture can be released using ESC.
+//!
+//! ```rust,no_run
+//! use ashpd::desktop::input_capture::{Barrier, Capabilities, InputCapture};
+//! use futures_util::StreamExt;
+//! use reis::{
+//!     ei::{self, keyboard::KeyState},
+//!     event::{DeviceCapability, EiEvent, KeyboardKey},
+//!     tokio::{EiConvertEventStream, EiEventStream},
+//! };
+//! use std::{collections::HashMap, os::unix::net::UnixStream, sync::OnceLock, time::Duration};
+//!
+//! #[allow(unused)]
+//! enum Position {
+//!     Left,
+//!     Right,
+//!     Top,
+//!     Bottom,
+//! }
+//!
+//! static INTERFACES: OnceLock<HashMap<&'static str, u32>> = OnceLock::new();
+//!
+//! async fn run() -> ashpd::Result<()> {
+//!     let input_capture = InputCapture::new().await.unwrap();
+//!
+//!     let (session, _cap) = input_capture
+//!         .create_session(
+//!             &ashpd::WindowIdentifier::default(),
+//!             Capabilities::Keyboard | Capabilities::Pointer | Capabilities::Touchscreen,
+//!         )
+//!         .await?;
+//!
+//!     // connect to eis server
+//!     eprintln!("connect_to_eis");
+//!     let fd = input_capture.connect_to_eis(&session).await?;
+//!
+//!     // create unix stream from fd
+//!     let stream = UnixStream::from(fd);
+//!     stream.set_nonblocking(true)?;
+//!
+//!     // create ei context
+//!     let context = ei::Context::new(stream)?;
+//!     context.flush().unwrap();
+//!
+//!     let mut event_stream = EiEventStream::new(context.clone())?;
+//!     let interfaces = INTERFACES.get_or_init(|| {
+//!         let mut m = HashMap::new();
+//!         m.insert("ei_connection", 1);
+//!         m.insert("ei_callback", 1);
+//!         m.insert("ei_pingpong", 1);
+//!         m.insert("ei_seat", 1);
+//!         m.insert("ei_device", 2);
+//!         m.insert("ei_pointer", 1);
+//!         m.insert("ei_pointer_absolute", 1);
+//!         m.insert("ei_scroll", 1);
+//!         m.insert("ei_button", 1);
+//!         m.insert("ei_keyboard", 1);
+//!         m.insert("ei_touchscreen", 1);
+//!         m
```

Comment:
Maybe using `HashMap::from(&[])` would make the code slightly nicer?

Reply:
done